protobuf/src/test/java/io/grpc/protobuf/ProtoUtilsTest.java

```diff
@@ -60,6 +64,24 @@ public void testJsonRoundtrip() throws Exception {
     assertEquals(proto, marshaller.parse(is));
   }
 
+  @Test
+  public void testJsonMarshallerFailToPrint() throws Exception {
+    JsonFormat.Printer mockPrinter = mock(JsonFormat.Printer.class);
```

Comment:
Any reason to not use the real printer, and just pass in an invalid proto?

Reply:
I thought that mock provides the "resilience" to unit tests, but in here it doesn't.
It is my fault. I will fix it as below, do you think?

``` diff
diff --git a/protobuf/src/test/java/io/grpc/protobuf/ProtoUtilsTest.java b/protobuf/src/test/java/io/grpc/protobuf/ProtoUtilsTest.java
index 5e9ccaf..3910d27 100644
--- a/protobuf/src/test/java/io/grpc/protobuf/ProtoUtilsTest.java
+++ b/protobuf/src/test/java/io/grpc/protobuf/ProtoUtilsTest.java
@@ -20,10 +20,10 @@ import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.fail;
 import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;

 import com.google.common.io.ByteStreams;
-import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Any;
+import com.google.protobuf.ByteString;
 import com.google.protobuf.Type;
 import com.google.protobuf.util.JsonFormat;
 import io.grpc.MethodDescriptor.Marshaller;
@@ -66,16 +66,15 @@ public class ProtoUtilsTest {

   @Test
   public void testJsonMarshallerFailToPrint() throws Exception {
-    JsonFormat.Printer mockPrinter = mock(JsonFormat.Printer.class);
-    when(mockPrinter.print(proto)).thenThrow(new InvalidProtocolBufferException(""));
+    Any nonPrintableProto = Any.newBuilder().setValue(ByteString.copyFromUtf8("invalid")).build();

-    Marshaller<Type> marshaller = ProtoUtils.jsonMarshaller(
-            Type.getDefaultInstance(),
+    Marshaller<Any> marshaller = ProtoUtils.jsonMarshaller(
+            nonPrintableProto,
             mock(JsonFormat.Parser.class),
-            mockPrinter
+            JsonFormat.printer()
     );
     try {
-      marshaller.stream(proto);
+      marshaller.stream(nonPrintableProto);
       fail("Expected exception");
     } catch (StatusRuntimeException e) {
       assertEquals(Status.Code.INTERNAL, e.getStatus().getCode());
```

Reply:
That's probably fine.  I would have personally just tried to decode the byte 0xFF, but you can do it that way too.

Reply:
I have pushed a rebased commit like above.

I have tried to decode the byte `0xFF`, but it fail at parsing phase.
``` java
Type.parseFrom(new byte[] {(byte) 0xFF});
```
So, what do you mean "decode the byte 0xFF"? I think I might have misunderstanding.

Reply:
That's correct, I think that's more clear than mocking.

Reply:
OK, I have tried but I got fail parsing phase.

``` java
  @Test
  public void testJsonMarshallerFailToPrint() throws Exception {
    try {
      Marshaller<Type> marshaller = ProtoUtils.jsonMarshaller(Type.getDefaultInstance());
      marshaller.stream(Type.parseFrom(new byte[] { (byte)0xFF }));
      fail("Expected exception");
    } catch (StatusRuntimeException e) {
      assertEquals(Status.Code.INTERNAL, e.getStatus().getCode());
    }
  }
```

```
io.grpc.protobuf.ProtoUtilsTest > testJsonMarshallerFailToPrint FAILED
    com.google.protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
        at com.google.protobuf.InvalidProtocolBufferException.truncatedMessage(InvalidProtocolBufferException.java:86)
        at com.google.protobuf.CodedInputStream$ArrayDecoder.readRawByte(CodedInputStream.java:1261)
        at com.google.protobuf.CodedInputStream$ArrayDecoder.readRawVarint64SlowPath(CodedInputStream.java:1149)
        at com.google.protobuf.CodedInputStream$ArrayDecoder.readRawVarint32(CodedInputStream.java:1043)
        at com.google.protobuf.CodedInputStream$ArrayDecoder.readTag(CodedInputStream.java:642)
        at com.google.protobuf.Type.<init>(Type.java:49)
        at com.google.protobuf.Type.<init>(Type.java:13)
        at com.google.protobuf.Type$1.parsePartialFrom(Type.java:1971)
        at com.google.protobuf.Type$1.parsePartialFrom(Type.java:1966)
        at com.google.protobuf.AbstractParser.parsePartialFrom(AbstractParser.java:163)
        at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:197)
        at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:209)
        at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:214)
        at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:49)
        at com.google.protobuf.Type.parseFrom(Type.java:556)
        at io.grpc.protobuf.ProtoUtilsTest.testJsonMarshallerFailToPrint(ProtoUtilsTest.java:73)

21 tests completed, 1 failed, 1 skipped
```

This works.
```
  @Test
  public void testJsonMarshallerFailToPrint() {
    Marshaller<Any> marshaller = ProtoUtils.jsonMarshaller(Any.getDefaultInstance());
    try {
      marshaller.stream(Any.newBuilder().setValue(ByteString.copyFromUtf8("invalid")).build());
      fail("Expected exception");
    } catch (StatusRuntimeException e) {
      assertEquals(Status.Code.INTERNAL, e.getStatus().getCode());
    }
  }
```